pkg/node-port-handler/node-port-handler.go

```diff
@@ -0,0 +1,120 @@
+package node_port_handler
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"net/http"
+	"time"
+)
+
+const vmVirtualIP = "192.168.127.2"
+
+func NodePortHandler(clientset *kubernetes.Clientset) cache.SharedIndexInformer {
+	factory := informers.NewSharedInformerFactory(clientset, 5*time.Minute)
+	informer := factory.Core().V1().Services().Informer()
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			service := obj.(*v1.Service)
+			log.Infof("Added service '%s' of type '%s'", service.GetName(), service.Spec.Type)
+			if service.Spec.Type != v1.ServiceTypeNodePort {
+				log.Debugf("Service '%s' is not of type NodePort. Nothing to do.", service.GetName())
+				return
+			}
+
+			for _, port := range service.Spec.Ports {
+				expose(port.NodePort)
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldService := oldObj.(*v1.Service)
+			newService := newObj.(*v1.Service)
+
+			log.Infof("Updated service '%s'", newService.GetName())
+			if oldService.Spec.Type == v1.ServiceTypeNodePort && newService.Spec.Type != v1.ServiceTypeNodePort {
```

Comment:
I am not sure if the case distinctions are the best way to do it. Just exposing/unexposing all the ports on each change to a NodePort service would simplify the code but would potentially also lead to unnecessary requests. What do you think?

Reply:
It seems to me that the code below will do exactly as much work as your current code while being shorter
```
if oldService.Spec.Type == v1.ServiceTypeNodePort {
  // unexpose all oldService ports
}
if newService.Spec.Type == v1.ServiceTypeNodePort {
  // expose all newService ports
}
```

If we want to limit some requests, we could build a list of ports to unexpose, build a list of ports to expose, and don't do anything for ports which are present in both lists. For now I think it's fine to just expose/unexpose everything.